pkg/validation/attributes/attributes_test.go

```diff
@@ -0,0 +1,257 @@
+package attributes
+
+import (
+	"io/ioutil"
+	"testing"
+
+	"github.com/devfile/api/v2/pkg/apis/workspaces/v1alpha2"
+	apiAttributes "github.com/devfile/api/v2/pkg/attributes"
+	"github.com/stretchr/testify/assert"
+	"sigs.k8s.io/yaml"
+)
+
+func TestValidateGlobalAttributeBasic(t *testing.T) {
+
+	tests := []struct {
+		name     string
+		testFile string
+		expected v1alpha2.DevWorkspaceTemplateSpec
+		wantErr  bool
+	}{
+		{
+			name:     "Successful global attribute substitution",
+			testFile: "test-fixtures/all/devfile-good.yaml",
+			expected: v1alpha2.DevWorkspaceTemplateSpec{
```

Comment:
For these tests in general, consider putting all input/output in files -- it's hard to track where replacements have occurred between a yaml file and a golang struct initialization.

Reply:
Addressed